lib/hw1/JavaScript homework.ipynb

```diff
@@ -90,12 +98,646 @@
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "Teacher { _name: 'Ion', _birthDate: 1621791750033, _employeeId: 69 }"
+      ]
+     },
+     "execution_count": 7,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "demoted_promoted_teacher = promoted_teacher.demote()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "Administrator {\n",
+       "  _name: 'unknown',\n",
+       "  _birthDate: 1621791755825,\n",
+       "  _employeeId: 0\n",
+       "}"
+      ]
+     },
+     "execution_count": 8,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "admin = new Administrator()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "'JOBS DONE'"
+      ]
+     },
+     "execution_count": 9,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "admin.do_administrator_stuff()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "ProDean { _name: 'Ioana', _birthDate: 1621791762498, _employeeId: 42 }"
+      ]
+     },
+     "execution_count": 10,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "prodean = new ProDean('Ioana', Date.now(), 42);"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "Dean { _name: 'Ioana', _birthDate: 1621791762498, _employeeId: 42 }"
+      ]
+     },
+     "execution_count": 11,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "promoted_prodean = prodean.promote()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 12,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "ProDean { _name: 'Ioana', _birthDate: 1621791762498, _employeeId: 42 }"
+      ]
+     },
+     "execution_count": 12,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "demoted_promoted_prodean = promoted_prodean.demote()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "[Function (anonymous)]"
+      ]
+     },
+     "execution_count": 1,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "function Person(name='unknown', birthDate=Date.now()) {\n",
+    "    this._name = name;\n",
+    "    this._birthDate = birthDate;\n",
+    "}\n",
+    "\n",
+    "function Student(name, birthDate) {\n",
```

Comment:
This is kinda incomplete. You are just mounting the properties from the constructor, not actually applying a prototypal inheritance.

Remember if we have Child inheriting Parent, then an instance of Child will look like this:

```
instance -prototype-> Child.prototype -prototype-> Parent.prototype
```

Where `.prototype` is a property named `prototype` set on the 'class' and `-prototype->` is the actual prototype of the object.

So the prototype of Child.prototype must be Parent.prototype.

Reply:
Therefore running

```
new Student() instanceof Person
```
yields false because the link between Student.prototype and Person.prototype is not active